contracts/dtoken/src/ft.rs

```diff
@@ -44,7 +44,38 @@ impl FungibleTokenReceiver for Contract {
         let action: &str = msg["action"].as_str().unwrap();
         match action {
             "SUPPLY" => self.supply(amount),
-            "REPAY" => self.repay(amount),
+            "REPAY" => self.repay(amount, None),
+            "LIQUIDATION" => {
+                let memo_data = msg["memo"].clone();
+
+                let borrower =
+                    AccountId::new_unchecked(String::from(memo_data["borrower"].as_str().unwrap()));
+                let borrowing_dtoken = AccountId::new_unchecked(String::from(
+                    memo_data["borrowing_dtoken"].as_str().unwrap(),
+                ));
+                let liquidator = AccountId::new_unchecked(String::from(
+                    memo_data["liquidator"].as_str().unwrap(),
+                ));
+                let collateral_dtoken = AccountId::new_unchecked(String::from(
+                    memo_data["collateral_dtoken"].as_str().unwrap(),
+                ));
+                let liquidation_amount = WBalance::from(
+                    memo_data["liquidation_amount"]
+                        .as_str()
+                        .unwrap()
+                        .parse::<u128>()
+                        .unwrap(),
+                );
+
+                self.liquidation(
+                    borrower,
+                    borrowing_dtoken,
+                    liquidator,
+                    collateral_dtoken,
+                    liquidation_amount,
+                );
+                PromiseOrValue::Value(amount)
```

Comment:
@convowork1 when you return `PromiseOrValue::Value(amount)` from `ft_on_transfer` you notifiy FungibleTokens that you have not used `amount` of transferred tokens so they should be returned to the user.

please check behavior of `ft_transfer_call` https://www.notion.so/dehive/Near-4d89f4d98ee345fb81fbc4eb0bd4d896

Instead of immediate return some value from `ft_on_transfer` flow should be built in such way so that success value returned only at the end of the flow and in case of fails full (or partial) refund should be returned (aka `PromiseOrValue::Value(amount)` or `PromiseOrValue::Value(amount - spent_tokens)` )